registrations/utils.py

```diff
@@ -1,6 +1,31 @@
 import phonenumbers
+from django.conf import settings
+from django.core.cache import caches
+from temba_client.v2 import TembaClient
+
+CACHE = caches["contacts"]
+CACHE_EXPIRY_SECONDS = 600
 
 
 def normalise_msisdn(msisdn):
     msisdn = phonenumbers.parse(msisdn, "ZA")
     return phonenumbers.format_number(msisdn, phonenumbers.PhoneNumberFormat.E164)
+
+
+def get_contact(msisdn):
+    contact = CACHE.get(msisdn)
+    if not contact:
+        client = TembaClient(settings.RAPIDPRO_URL, settings.RAPIDPRO_TOKEN)
+        contact = client.get_contacts(urn="tel:%s" % msisdn).first()
+        CACHE.set(msisdn, contact, CACHE_EXPIRY_SECONDS)
```

Comment:
I'm struggling to see the use cases where we'll be fetching the contact from the cache. We have to remove it from the cache at the end of registration, since we'll be updating the contact at the end of the registration, so maybe it can just be put in the session instead? (Since we'll probably need to know if there's an existing contact, or if we need to create one, at the end of the registration)

Reply:
tbh I also thought that would be better but I couldn't figure out how to access the request session from the form where I wanted to use it. Any pointers for that?

Reply:
I'm fairly sure that we have the request in the form, so we should be able to access the session through that, ie. `self.request.session`

Reply:
Actually, I'm wrong on that. To get access to the request you'll have to pass it down to the form. You can probably use `get_form_kwargs` in the view to pass the request, and then in the `__init__` of the form pop it from the kwargs and store it on the form instance

Reply:
ok. I'll give that a try and remove the caching stuff if I can get it working. Thanks